pkg/flow/adapter/synchronizer/adapter.go

```diff
@@ -0,0 +1,147 @@
+		logger.Panic("Cannot create an instance of Correlation Key: %v", err)
+	}
+
+	return &adapter{
+		ceClient:       ceClient,
+		logger:         logger,
+		correlationKey: key,
+
+		responseTimeout: timeout,
+		sessions:        newStorage(),
+		sinkURL:         env.Sink,
+	}
+}
+
+// Returns if stopCh is closed or Send() returns an error.
+func (a *adapter) Start(ctx context.Context) error {
+	a.logger.Info("Starting Synchronizer Adapter")
+	return a.ceClient.StartReceiver(ctx, a.dispatch)
+}
+
+func (a *adapter) dispatch(ctx context.Context, event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	a.logger.Debugf("Received the event: %s", event.String())
+
+	if correlationID, exists := a.correlationKey.Get(event); exists {
+		return a.serveResponse(ctx, correlationID, event)
+	}
+
+	correlationID := a.correlationKey.Set(&event)
+	return a.serveRequest(ctx, correlationID, event)
+}
+
+// serveRequest creates the session for the incoming events and blocks the client.
+func (a *adapter) serveRequest(ctx context.Context, correlationID string, event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	a.logger.Debugf("Handling request %q", correlationID)
+
+	respChan := a.sessions.add(correlationID)
+	defer a.sessions.delete(correlationID)
+
+	sendErr := make(chan error)
+	defer close(sendErr)
+
+	go func(errChan chan error) {
+		if res := a.ceClient.Send(cloudevents.ContextWithTarget(ctx, a.sinkURL), event); cloudevents.IsUndelivered(res) {
+			errChan <- res
+		}
+	}(sendErr)
+
+	a.logger.Debugf("Request forwarded to %q", a.sinkURL)
+
+	t := time.NewTimer(time.Duration(a.responseTimeout))
+	defer t.Stop()
+
+	a.logger.Debugf("Waiting response for %q", correlationID)
+
+	select {
+	case err := <-sendErr:
+		a.logger.Errorf("Unable to forward the request: %v", err)
+		return nil, cloudevents.NewHTTPResult(http.StatusBadGateway, "unable to forward the request: %v", err)
+	case result := <-respChan:
+		if result == nil {
+			a.logger.Errorf("No response from %q", correlationID)
+			return nil, cloudevents.NewHTTPResult(http.StatusInternalServerError, "failed to communicate the response")
+		}
+		a.logger.Debugf("Received response for %q", correlationID)
+		return result, cloudevents.ResultACK
+	case <-t.C:
+		a.logger.Errorf("Request %q timed out", correlationID)
+		return nil, cloudevents.NewHTTPResult(http.StatusGatewayTimeout, "backend did not respond in time")
+	}
+}
+
+// serveResponse matches event's correlation key and writes response back to the session's communication channel.
+func (a *adapter) serveResponse(ctx context.Context, correlationID string, event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	a.logger.Debugf("Handling response %q", correlationID)
+
+	responseChan, exists := a.sessions.open(correlationID)
+	if !exists {
+		a.logger.Errorf("Session for %q does not exist", correlationID)
+		return nil, cloudevents.NewHTTPResult(http.StatusBadGateway, "client session does not exist")
```

Comment:
is `BadGateway` the intended code to return here? I think this could lead to confusion.

Reply:
Yes, it was intended. What would be the better status here? Please note the context of this error - it is a response from the abstract backend to the synchronizer, when the client session does not exist (timeout, user's Ctrl+C, pod restart, etc.)